Add ability to hide KubeObjectMenu Edit and Remove buttons in extensions

Nokel81 opened the pull request:
- The core of the solution is to add event listener handlers for the `onContextMenuOpen` event for `KubeObject` in a similar fashion to that of catalog entities. This allows any extension to add/remove items and means that the producers don't have to worry about things like `<Menu>` and `<MenuItem>` which the old `kubeObjectMenuItems` registration pattern on `LensRendererExtension` requires.

Signed-off-by: Sebastian Malton <sebastian@malton.name>

fixes #5105

aleksfront reviewed (commented):
Should we set default `true` values for brand new `showEditButton` & `showRemoveButton` fields to prevent breaking changes? I guess all working extensions showing those button by default?

ixrock commented on src/renderer/components/kube-object-menu/kube-object-menu.tsx:
```diff
@@ -41,11 +41,11 @@ class NonInjectedKubeObjectMenu<TKubeObject extends KubeObject> extends React.Co
 
   get isEditable() {
-    return this.props.editable ?? Boolean(this.store?.patch);
+    return this.props.editable ?? this.store?.showEditButton ?? false;
```
Having `showEditButton` and other field on `KubeObjectStore` just weird as heck.
Why only `props.editable` not enough to control this items?

Nokel81 replied:
I agree it is extremely weird.

Nokel81 replied:
It is not enough because the problem here is the fact that the extension API does not have a mechanism to remove these buttons for any kube objects, let alone for new kube objects which an extension might actually control.

Another solution would be to require extension developers to add these buttons explicitly but that is a breaking change.

jakolehm commented on src/common/k8s-api/kube-object.store.ts:
```diff
@@ -64,6 +64,24 @@ export abstract class KubeObjectStore<T extends KubeObject> extends ItemStore<T>
+   * This should only be set once, in the constructor of base classes since the views will
+   * not be reactive to any changes to this value.
+   */
+  public readonly showEditButton = true;
```
IMHO it's not good practice to mix UI logic to store/api level. Maybe these could be called as `isEditable` & `isRemovable` ?

Nokel81 replied:
I agree that we should not mix them. But I don't see exactly how changing the name would help.

Maybe we could just extract this into some sort of extension API call which can be called to turn it off for any store. WDYT?

jakolehm replied:
`showEditButton` says that it affects the edit button visibility. `isEditable` is much more generic, edit button can use this information but so can other functionality. Small detail but I think we should use more generic names if possible.

Nokel81 replied:
Okay will change.

Nokel81 committed:
Allow extensions to iteract with kube detail menu items like CatalogEntity menu items

- Allow registering handlers

- Cannot remove menu items

- Can add menu items in any location

- Add example of StatefulSets

- Could be used to deprecate the old API

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
fix lint

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
fix test

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
update snapshots

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
try and fix tests (again)

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
try and fix test again

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
better fix

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
readd awaits

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Remove using <Animated /> to help with determinism

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Make <MenuActions /> not animated if toolbar to increase determinism

Signed-off-by: Sebastian Malton <sebastian@malton.name>

jakolehm reviewed (commented):
Quite a big diff but probably all changes have a good reason. Please update the PR description to match the implementation (and maybe add why so many changes were required).

Nokel81 commented on src/common/catalog/catalog-entity.ts:
```diff
@@ -278,11 +278,22 @@ export interface CatalogEntitySettingsMenu {
   };
 }
 
+export interface CatalogEntityContextMenuNavigate {
```
This just clarifies that this is actually (previous to this PR) supported here witch some additional documentation. 

This probably could be moved into a separate PR if desired.

Nokel81 commented on src/extensions/lens-renderer-extension.ts:
```diff
@@ -49,6 +50,7 @@ export class LensRendererExtension extends LensExtension {
   topBarItems: TopBarRegistration[] = [];
   additionalCategoryColumns: AdditionalCategoryColumnRegistration[] = [];
   customCategoryViews: CustomCategoryViewRegistration[] = [];
+  kubeObjectHandlers: KubeObjectHandlerRegistration[] = [];
```
This is how the event listeners are registered

Nokel81 commented on src/extensions/renderer-api/components.ts:
```diff
@@ -43,14 +47,23 @@ export type {
+  ConfirmDialogParams,
+  ConfirmDialogProps,
+};
+export const ConfirmDialog = Object.assign(_ConfirmDialog, {
```
It was required to transform `ConfirmDialog` to use dependency injection. This was required because confirm dialog was not cooperating with behaviour unit tests (it had global shared state).

Because this transformation was required, I needed to move the `open` and `confirm` functions away from being statics on the component to be there own injectables. This rectifies that for the extension API only.

Nokel81 commented on src/renderer/catalog/normalize-menu-item.injectable.ts:
```diff
@@ -0,0 +1,43 @@
+
+export type NormalizeCatalogEntityContextMenu = (menuItem: CatalogEntityContextMenu) => NormalizedCatalogEntityContextMenu;
+
+const normalizeCatalogEntityContextMenuInjectable = getInjectable({
```
Because `ConfirmDialog` needed to be DI-ed, I took the opportunity to make this code more DRY. It was repeated 3 or 4 times throughout the codebase previously.

Nokel81 commented on src/renderer/components/+extensions/attempt-install-by-info.injectable.tsx:
```diff
@@ -0,0 +1,138 @@
+/**
```
Even git makes this code look like it is completely new, it is not. It is mostly just the renaming of the file previously at `+extensions/attempt-install-by-info/attempt-install-by-info.tsx`, just with the `getInjectable` call added, the factory not exported, and `confirm` (which is not injectable) added as a dependency.

Nokel81 commented on src/renderer/components/+extensions/confirm-uninstall-extension.injectable.tsx:
```diff
@@ -0,0 +1,54 @@
+/**
```
This file's diff is similar to that of the `attempt-install-by-info` ones

Nokel81 commented on src/renderer/components/animate/animate.tsx:
```diff
@@ -46,15 +46,24 @@ export class Animate extends React.Component<AnimateProps> {
     return React.Children.only(this.props.children) as React.ReactElement<React.HTMLAttributes<any>>;
   }
 
+  private toggle(enter: boolean) {
```
I made this change while debugging why `Menu` and `Drawer` were being non-deterministic in behavioural unit test. This change doesn't change the behaviour and ultimately I just decided to remove the use of `<Animate>` from the tests.

The other option would be to make `requestAnimateFrame` (which has side effects) a dependency and override it in tests. Could be very useful for future work.

Nokel81 commented on src/renderer/components/confirm-dialog/with-confirm.injectable.ts:
```diff
@@ -0,0 +1,21 @@
+import { getInjectable } from "@ogre-tools/injectable";
+import openConfirmDialogInjectable from "./open.injectable";
+
+export type WithConfirmation = (params: ConfirmDialogParams) => () => void;
```
This is a useful higher order function which can better express what the intent is. Namely, that I want to confirm something as a callback instead of immediately.

Nokel81 commented on src/renderer/components/drawer/drawer.tsx:
```diff
@@ -199,7 +199,7 @@ class NonInjectedDrawer extends React.Component<DrawerProps & Dependencies, Stat
               </div>
               {toolbar}
-              <Icon material="close" onClick={this.close}/>
+              <Icon material="close" tooltip="Close" onClick={this.close}/>
```
I noticed this bug while testing, it is the only icon for drawers that doesn't have a tooltip.

This could be moved to a separate PR.

Nokel81 commented on src/renderer/components/icon/icon.tsx:
```diff
@@ -15,22 +15,69 @@ import { withTooltip } from "../tooltip";
-  active?: boolean;           // apply active-state styles
-  interactive?: boolean;      // indicates that icon is interactive and highlight it on focus/hover
-  focusable?: boolean;        // allow focus to the icon + show .active styles (default: "true", when icon is interactive)
+export interface BaseIconProps {
```
I separated out these props so that I could export this subset as part of the `KubeObjectHandlerRegistration` type.

The bonus of using doc comments it that our API docs should be better documented now.

Nokel81 commented on src/renderer/components/kube-object-menu/dependencies/static-kube-object-menu-items.ts:
```diff
@@ -37,11 +36,4 @@ export const staticKubeObjectMenuItems = [
       MenuItem: ReplicaSetMenu,
     },
   },
-  {
```
I changed this single `KubeObjectMenu` registration into a `KubeObjectHandler` registration to show in this PR how it can be done.

Nokel81 commented on src/renderer/components/layout/sidebar-cluster.tsx:
```diff
@@ -86,8 +64,17 @@ const NonInjectedSidebarCluster = observer(({ clusterEntity, hotbarStore }: Depe
+    menuItems.replace([{ title, onClick }]);
+    clusterEntity.onContextMenuOpen({
+      menuItems,
+      navigate: (url, forceMainFrame = true) => {
```
This is copied from line 25 in the left hand side of the diff. And is why I updated the typings above.

ixrock commented on src/renderer/components/+extensions/attempt-install-by-info.injectable.tsx:
```diff
@@ -0,0 +1,138 @@
+    const { name, version, requireConfirmation = false } = info;
+    const disposer = extensionInstallationStateStore.startPreInstall();
+    const baseUrl = await getBaseRegistryUrl();
+    const registryUrl = new URLParse(baseUrl).set("pathname", name).toString();
```
Why not using standard `new URL()`? (it's also available in `nodejs`)

```ts
const registryUrl = new URL(baseUrl);
registryUrl.pathname = name;
registryUrl.toString(); // result
```

Nokel81 replied:
Mostly habit, but fair point, and given that they now recommend doing that we probably should as well. I will make a follow up PR to remove the `url-parse` library from our repo.

ixrock commented on src/common/catalog/catalog-entity.ts:
```diff
@@ -278,11 +278,22 @@ export interface CatalogEntitySettingsMenu {
+  /**
+   * @param pathname The location to navigate to in the main iframe
+   */
+  (pathname: string, forceMainFrame?: boolean): void;
+  /**
+   * @param pathname The location to navigate to in the current iframe. Useful for when called within the cluster frame
+   */
+  (pathname: string, forceMainFrame: false): void;
```
A bit weird definition. Would be more logical imho if by default with `false` or `undefined` second argument it will use current iframe, and force main iframe only with optional `forceMainFrame: true`. Wdyt?

Nokel81 replied:
This documentation was done to match the current implementation, yeah the default is a bit confusing.

ixrock commented on src/renderer/components/confirm-dialog/confirm-dialog.tsx:
```diff
@@ -124,10 +114,10 @@ export class ConfirmDialog extends React.Component<ConfirmDialogProps> {
         onClose={this.onClose}
         close={this.close}
-        {...(dialogState.isOpen ? { "data-testid":"confirmation-dialog" } : {})}
+        {...(isOpen ? { "data-testid": "confirmation-dialog" } : {})}
```
Why `data-testid` passed only when dialog is open?

Nokel81 replied:
I questioned this myself, it is so that while the dialog is closed `findByTestId` won't find it.

ixrock replied:
In general we should not render dialog itself, instead of hiding reachability in tests. 
But this is fine for now I guess.

ixrock commented on src/renderer/components/layout/sidebar-cluster.tsx:
```diff
@@ -7,74 +7,52 @@ import styles from "./sidebar-cluster.module.scss";
+      <div className={styles.SidebarCluster}>
+        <Avatar
+          title="??"
+          background="var(--halfGray)"
```
Why not adding background directly to `styles.loadingAvatar`?

Nokel81 replied:
Because inline styles (which is what `<Avatar>` uses for `background`, `width`, and `height`) overrides class based styling, and background is something is always set by code.

ixrock replied:
We could use `background: var(--halfGray) !important;` (let's try next time avoid using css-variables inside components as possible == easier refactoring at later point / separation of concerns)

ixrock commented on src/renderer/components/menu/menu.tsx:
```diff
@@ -319,28 +321,40 @@ export class Menu extends React.Component<MenuProps, State> {
+        ref={this.bindRef}
+        className={className}
+        style={{
+          left: this.state?.menuStyle?.left,
+          top: this.state?.menuStyle?.top,
```
`this.state` might not exists?

Nokel81 replied:
Nope, just copy pasted it the old code. Probably should be cleaned up in the future.
